gtsam/slam/EssentialMatrixConstraint.cpp

```diff
@@ -43,16 +43,20 @@ bool EssentialMatrixConstraint::equals(const NonlinearFactor& expected,
 
 /* ************************************************************************* */
 Vector EssentialMatrixConstraint::evaluateError(const Pose3& p1,
-    const Pose3& p2, boost::optional<Matrix&> Hp1,
-    boost::optional<Matrix&> Hp2) const {
+    const Pose3& p2, OptionalMatrixType Hp1,
+    OptionalMatrixType Hp2) const {
 
   // compute relative Pose3 between p1 and p2
   Pose3 _1P2_ = p1.between(p2, Hp1, Hp2);
 
   // convert to EssentialMatrix
   Matrix D_hx_1P2;
-  EssentialMatrix hx = EssentialMatrix::FromPose3(_1P2_,
-      (Hp1 || Hp2) ? boost::optional<Matrix&>(D_hx_1P2) : boost::none);
+  EssentialMatrix hx;
+  if (Hp1 || Hp2) {
+    hx = EssentialMatrix::FromPose3(_1P2_, D_hx_1P2);
+  } else {
+      hx = EssentialMatrix::FromPose3(_1P2_, OptionalNone);
+  }
```

Comment:
Why did this have to change?  Could this stay as
```c++
EssentialMatrix hx = EssentialMatrix::FromPose3(_1P2_, (Hp1 || Hp2) ? &D_hx_1P2 : 0);
```
?

Reply:
Yes it can, I think I changed it because I found it confusing. I will revert it.

Reply:
Ah I changed it because `OptionalNone` and `D_hx_1P2` are of different types.

Reply:
Well, could you instead use nullptr?

Reply:
No nullptr and D_hx_1P2 (matrix) are different types too. They can’t be used with a ternary operator. OptionalNone typecasts to nullptr

Reply:
Check the original code, it’s a pointer. This idiom is used all over GTSAM so I’m a bit worried it’s broken.

Reply:
I'll resolve - might fix in follow up PR